frontend/src/metabase/visualizations/components/ScalarValue.styled.tsx

```diff
@@ -11,3 +22,94 @@ export const ScalarRoot = styled.div`
   width: 100%;
   height: 100%;
 `;
+
+export const ScalarValueWrapper = styled.h1<ScalarValueProps>`
+  cursor: pointer;
+  &:hover {
+    color: ${color("brand")};
+  }
+
+  font-size: ${props => {
+    const MIN_SIZE_REM = 2.2;
+    const MAX_SIZE_REM = 12;
+
+    const fontSizeAdjustment = computeFontSizeAdjustment(props);
+
+    // fail-safe to make sure the font size is within a reasonable range
+    const clampedScaledFontSize = Math.min(
+      Math.max(MIN_SIZE_REM + fontSizeAdjustment, MIN_SIZE_REM),
+      MAX_SIZE_REM,
+    );
+
+    return `${clampedScaledFontSize}rem`;
+  }};
+`;
+
+export const WIDTH_ADJUSTMENT_FACTOR = 4;
+export const HEIGHT_ADJUSTMENT_FACTOR = 4;
+export const MAX_HEIGHT_GRID_SIZE = 10;
+
+export function computeFontSizeAdjustment({
+  isDashboard,
+  gridSize,
+  minGridSize,
+  width: widthPx,
+  height: heightPx,
+  totalNumGridCols,
+}: ScalarValueProps): number {
+  if (
+    !isDashboard ||
+    !gridSize ||
+    !minGridSize ||
+    !widthPx ||
+    !heightPx ||
+    !totalNumGridCols
+  ) {
+    return 0;
+  }
+
+  const { height: gridSizeHeight, width: gridSizeWidth } = gridSize;
+  const { height: minGridSizeHeight, width: minGridSizeWidth } = minGridSize;
+  if (
+    !gridSizeHeight ||
+    !gridSizeWidth ||
+    !minGridSizeHeight ||
+    !minGridSizeWidth
+  ) {
+    return 0;
+  }
+
+  // at small viewport widths totalNumGridCols is set to 1, but the dashcard's gridSize.width isn't updated.
+  // in that scenario, the dashcard's grid width should be treated as 1 as well because it spans the entire grid
+  const dashCardGridWidth = Math.min(totalNumGridCols, gridSizeWidth);
+  const widthPxPerGridUnit = widthPx / dashCardGridWidth;
+  const maxWidthPx = totalNumGridCols * widthPxPerGridUnit;
+  // 3 is taken from Scalar's min grid size -- should make it a constant.
+  const minWidthPx = minGridSizeWidth * widthPxPerGridUnit;
```

Comment:
nit: Improve naming somehow to make it clear what object values belong to. E.g. `minGridWidthPx`.